beep/structure/base.py

```diff
@@ -0,0 +1,1373 @@
+                the voltage interpolation range endpoints.
+            resolution (int): resolution of interpolated data.
+            step_type (str): which step to interpolate i.e. 'charge' or 'discharge'
+            reg_cycles (list): list containing cycle indicies of regular cycles
+            axis (str): which column to use for interpolation
+            desc (str): Description to print to tqdm column.
+
+        Returns:
+            pandas.DataFrame: DataFrame corresponding to interpolated values.
+        """
+
+        if not desc:
+            desc = \
+                f"Interpolating {step_type} ({v_range[0]} - {v_range[1]})V " \
+                f"({resolution} points)"
+
+        if step_type == "discharge":
+            step_filter = step_is_dchg
+        elif step_type == "charge":
+            step_filter = step_is_chg
+        else:
+            raise ValueError("{} is not a recognized step type")
+        incl_columns = [
+            "test_time",
+            "voltage",
+            "current",
+            "charge_capacity",
+            "discharge_capacity",
+            "charge_energy",
+            "discharge_energy",
+            "internal_resistance",
+            "temperature",
+        ]
+        all_dfs = []
+        cycle_indices = self.raw_data.cycle_index.unique()
+        cycle_indices = sorted([c for c in cycle_indices if c in reg_cycles])
+
+        for cycle_index in tqdm(cycle_indices, desc=desc):
+            # Use a cycle_index mask instead of a global groupby to save memory
+            new_df = (
+                self.raw_data.loc[self.raw_data["cycle_index"] == cycle_index]
+                    .groupby("step_index")
+                    .filter(step_filter)
+            )
+            if new_df.size == 0:
+                continue
+
+            if axis in ["charge_capacity", "discharge_capacity"]:
+                axis_range = [self.raw_data[axis].min(),
+                              self.raw_data[axis].max()]
+                new_df = interpolate_df(
+                    new_df,
+                    axis,
+                    field_range=axis_range,
+                    columns=incl_columns,
+                    resolution=resolution,
+                )
+            elif axis == "test_time":
+                axis_range = [new_df[axis].min(), new_df[axis].max()]
+                new_df = interpolate_df(
+                    new_df,
+                    axis,
+                    field_range=axis_range,
+                    columns=incl_columns,
+                    resolution=resolution,
+                )
+            elif axis == "voltage":
+                new_df = interpolate_df(
+                    new_df,
+                    axis,
+                    field_range=v_range,
+                    columns=incl_columns,
+                    resolution=resolution,
+                )
+            else:
+                raise NotImplementedError
```

Comment:
Can we add an error message here to indicate if the axis is wrong?  Might help if e.g. someone mispells the axis parameter.

Reply:
addressed in [0a7fa78](https://github.com/TRI-AMDD/beep/pull/150/commits/0a7fa78f4d39d99469b39718e143d48ad8fe8826)